shared/lib-hoppy/hoppy/async_hoppy_client.py

```diff
@@ -0,0 +1,211 @@
+from hoppy.hoppy_properties import ExchangeProperties, QueueProperties
+from pika import BasicProperties
+
+MAX_RETRIES_REACHED = "Max retries reached"
+TIMED_OUT = "Request timed out"
+UNDECODABLE = "JSON response could not be parsed"
+UNEXPECTED_PUBLISH_ERROR = "Unexpected Error Caught While Publishing"
+
+
+class AsyncHoppyClient:
+
+    def __init__(self,
+                 name: str,
+                 app_id: str,
+                 config: dict,
+                 exchange_properties: ExchangeProperties,
+                 request_queue_properties: QueueProperties,
+                 reply_queue_properties: QueueProperties,
+                 request_routing_key: str,
+                 reply_routing_key: str,
+                 max_latency: int = 30,
+                 response_reject_and_requeue_attempts: int = 3):
+        self.name = name
+        self.app_id = app_id
+        self.config = config
+
+        self.exchange_properties = exchange_properties
+        self.request_queue_properties = request_queue_properties
+        self.reply_queue_properties = reply_queue_properties
+
+        self.max_latency = max_latency
+        self.response_reject_and_requeue_attempts = response_reject_and_requeue_attempts
+
+        self.async_publisher = AsyncPublisher(config=self.config,
+                                              exchange_properties=exchange_properties,
+                                              queue_properties=request_queue_properties,
+                                              routing_key=request_routing_key)
+
+        self.async_consumer = AsyncConsumer(config=self.config,
+                                            exchange_properties=exchange_properties,
+                                            queue_properties=reply_queue_properties,
+                                            routing_key=reply_routing_key,
+                                            reply_callback=self._on_reply)
+
+        self.responses = {}
+        self.rejected = {}
+
+    def start(self, loop):
+        self.async_publisher.connect(loop)
+        self.async_consumer.connect(loop)
+
+    def stop(self):
+        self.async_publisher.stop()
+        self.async_consumer.stop()
+
+    async def make_request(self, request_id, body):
+        correlation_id = str(uuid.uuid4())
+        logging.info(
+            f"event=requestStarted id={request_id} "
+            f"client={self.name} "
+            f"queue={self.request_queue_properties.name} "
+            f"correlation_id={correlation_id}")
+        self.responses[correlation_id] = None
+
+        self.publish(body, correlation_id, request_id)
+
+        wait_for_response_time = time.time()
+        while True:
+            response = self.responses.get(correlation_id)
+            if response is not None:
+                return self.handle_response(request_id, correlation_id, response)
+            else:
+                self.wait_for_response(correlation_id,
```

Comment:
This method is way more readable now! I would just suggest renaming this call to something other than `wait_for_response`, since it's really just checking whether the timeout has been reached, and all the actual waiting happens in the await call below.